python/ffsim/states/slater.py

```diff
@@ -0,0 +1,218 @@
+                n_a / norb
+            )
+        else:
+            sampled_configuration_b = _sample_spinless_direct(rdm_b, shots, rng)
+
+        sampled_configuration = np.concatenate(
+            (sampled_configuration_b, sampled_configuration_a), axis=1
+        )
+
+    return sampled_configuration
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, pos_array: np.ndarray, empty_orbitals: np.ndarray, marginal: float
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Generates the conditional and marginal probabilities for adding a particle
+    to the available empty orbitals.
+
+    This is a step of the autoregressive sampling, and uses Bayes's rule.
+
+    Args:
+        rdm: A Numpy array with the one-body reduced density matrix.
+        pos_array: A numpy array with the positions of the particles.
+        empty_orbitals: A numpy array with the empty orbitals that a new particle
+            may occupy. The sorted union of ``pos_array`` and ``empty_orbitals``
+            must be equal to ``numpy.arange(num_orbitals)``.
+        marginal: The marginal probability associated to having the particles
+            in the positions determined by ``pos_array``.
+    Returns:
+        A tuple of Numpy arrays. The first is the unormalized conditionals for
+        adding a particle to any of the empty orbitals specified in the input.
+        The second is the marginal corresponding to having the particles in the
+        position array and one extra in all possible empty orbitals. Both arrays
+        follow the same order as ``empty_orbitals``.
+
+    """
+
+    conditionals = np.zeros(len(empty_orbitals), dtype=complex)
+    marginals = np.zeros(len(empty_orbitals), dtype=complex)
+
+    for i, orbital in enumerate(empty_orbitals):
+        new_pos_array = np.append(pos_array, [orbital])
+        rest_rdm = rdm[new_pos_array, :]
+        rest_rdm = rest_rdm[:, new_pos_array]
+        marginals[i] = np.linalg.det(rest_rdm)
+        conditionals[i] = marginals[i] / marginal
+
+    return conditionals, marginals
+
+
+def _autoregressive_slater(
+    rdm: np.ndarray,
+    norb: int,
+    nelec: int,
+    seed: np.random.Generator | int | None = None,
+) -> np.ndarray:
+    """
+    Autoregressively sample positions of particles for a Slater-determinant wave
+    function using a determinantal point process.
+
+    Args:
+        rdm: A numpy array with the one-body reduced density matrix.
+        norb: Number of orbitals.
+        nelec: Number of electrons.
+        seed: Either a Numpy random generator, an integer seed for the random number
+            generator or ``None``.
+    Returns:
+        A numpy array with the position of the sampled electrons.
+    """
+    rng = np.random.default_rng(seed)
+
+    position = []
+    marginal = []
+    position.append(
+        rng.choice(np.arange(norb), size=1, p=np.real(np.diag(rdm) / nelec))[0]
+    )
+    marginal.append(np.real(np.diag(rdm) / nelec)[position[0]])
```

Comment:
For `position` and `marginal`, instead of initializing them to `[]` and appending a single item, just initialize them with the item, e.g. `[item]`.